lib/ansible/module_utils/compat/datetime.py

```diff
@@ -3,36 +3,51 @@
 
 from __future__ import annotations
 
-from ansible.module_utils.six import PY3
+import datetime as _datetime
+import typing as t
 
-import datetime
+from ansible.module_utils.common.warnings import deprecate
 
 
-if PY3:
-    UTC = datetime.timezone.utc
-else:
-    _ZERO = datetime.timedelta(0)
+_UTC = _datetime.timezone.utc
 
-    class _UTC(datetime.tzinfo):
-        __slots__ = ()
 
-        def utcoffset(self, dt):
-            return _ZERO
+def _utcfromtimestamp(timestamp: float) -> _datetime.datetime:
+    """Construct an aware UTC datetime from a POSIX timestamp."""
+    return _datetime.datetime.fromtimestamp(timestamp, _UTC)
 
-        def dst(self, dt):
-            return _ZERO
 
-        def tzname(self, dt):
-            return "UTC"
+def _utcnow() -> _datetime.datetime:
+    """Construct an aware UTC datetime from time.time()."""
+    return _datetime.datetime.now(_UTC)
 
-    UTC = _UTC()
 
+_deprecated_shims_map: dict[str, t.Callable[..., object] | _datetime.timezone] = {
+    'UTC': _UTC,
+    'utcfromtimestamp': _utcfromtimestamp,
+    'utcnow': _utcnow,
+}
 
-def utcfromtimestamp(timestamp):  # type: (float) -> datetime.datetime
-    """Construct an aware UTC datetime from a POSIX timestamp."""
-    return datetime.datetime.fromtimestamp(timestamp, UTC)
+__all__ = tuple(_deprecated_shims_map)
 
 
-def utcnow():  # type: () -> datetime.datetime
-    """Construct an aware UTC datetime from time.time()."""
-    return datetime.datetime.now(UTC)
+def __getattr__(importable_name: str) -> t.Callable[..., object] | _datetime.timezone:
+    """Inject import-time deprecation warnings.
+
+    Specifically, for ``UTC``, ``utcfromtimestamp()`` and ``utcnow()``.
+    """
+    try:
+        importable = _deprecated_shims_map[importable_name]
+    except KeyError as key_err:
+        raise AttributeError(
+            f'cannot import name {importable_name !r} '
+            f'has no attribute ({__file__ !s})',
+        ) from key_err
```

Comment:
This more accurately reflects the standard behavior:

```suggestion
        raise AttributeError(f"module {__name__!r} has no attribute {key_err}") from None
```

Reply:
I suppose… Although, this exception isn't actually surfaced anywhere. The end-users get a regular `ImportError`, I think.

Reply:
The `AttributeError` is exposed when accessing attributes instead of importing them, such as:

```python
from ansible.module_utils.compat import datetime

datetime.something
```